oddoor/models/oddoor_lock.py

```diff
@@ -0,0 +1,63 @@
+# Copyright 2019 Creu Blanca
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models, _
+
+
+class OddoorLock(models.Model):
+
+    _name = 'oddoor.lock'
+    _description = 'Oddoor Lock'
+
+    name = fields.Char()
+    description = fields.Text()
+    group_ids = fields.Many2many(
+        'oddoor.group',
+    )
+    active = fields.Boolean(default=True, required=True)
+
+    @api.multi
+    def check_access_unique_virtual_key(self, unique_virtual_key):
+        self.ensure_one()
+        key = self.env['oddoor.key'].search([
+            '|', ('expiration_date', '=', False),
+            ('expiration_date', '<', fields.Datetime.now()),
+            ('unique_virtual_key', '=', unique_virtual_key),
+        ], limit=1)
+        if not key:
+            return False
+        return self.check_access(key)
+
+    @api.multi
+    def check_access(self, key):
+        result = self._check_access(key)
+        self.create_action(key, result)
+        return result
+
+    def create_action(self, key, result):
+        return self.env['oddoor.key.action'].create(self._create_action_vals(
+            key, result
+        ))
+
+    def _create_action_vals(self, key, result):
+        return {
+            'lock_id': self.id,
+            'key_id': key.id,
+            'result': 'accepted' if result else 'refused'
+        }
+
+    @api.multi
+    def _check_access(self, key):
+        self.ensure_one()
+        if key.expiration_date and key.expiration_date < fields.Datetime.now():
```

Comment:
This if will never be True since `check_access_unique_virtual_key` already filters. Am I missing something?

Reply:
Good point, I will remove the first filter in order to make all the same